finality-aleph/src/sync/substrate/state_notifier.rs

```diff
@@ -0,0 +1,81 @@
+use std::fmt::{Display, Error as FmtError, Formatter};
+
+use aleph_primitives::BlockNumber;
+use futures::StreamExt;
+use sc_client_api::client::{FinalityNotifications, ImportNotifications};
+use sp_runtime::traits::{Block as BlockT, Header as SubstrateHeader};
+use tokio::select;
+
+use crate::sync::{substrate::BlockId, ChainStateNotification, ChainStateNotifier, Header};
+
+/// What can go wrong when waiting for next chain state notification.
+#[derive(Debug)]
+pub enum Error {
+    JustificationStream,
+    ImportStream,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use Error::*;
+        match self {
+            JustificationStream => {
+                write!(f, "finalization notification stream has ended")
+            }
+            ImportStream => {
+                write!(f, "import notification stream has ended")
+            }
+        }
+    }
+}
+
+/// Substrate specific implementation of `ChainStateNotifier`.
+pub struct SubstrateChainStateNotifier<H, B>
```

Comment:
You should only need the `B` here, and replace all instances of `H` anywhere by `B::Header`. This will add a slightly weird requirement that `B::Header::Number = BlockNumber` and this might be slightly annoying to express, but it shouldn't be that terrible?

Reply:
Nice, I like it